File: Difficulty: Medium/Count Inversions/count-inversions.cpp
//{ Driver Code Starts
#include <bits/stdc++.h>
using namespace std;


// } Driver Code Ends

class Solution {
  public:
    int merge(vector<int> &arr, int low, int mid, int high){
        vector<int> temp;
        int left = low;
        int right = mid + 1;
        int count = 0;
        
        while(left <= mid && right <= high){
            if(arr[left] <= arr[right]){
                temp.push_back(arr[left]);
                left++;
            }else{
                temp.push_back(arr[right]);
                count+=(mid - left + 1);
                right++;
            }
        }
        
        while(left <= mid){
            temp.push_back(arr[left]);
            left++;
        }
        
        while(right <= high){
            temp.push_back(arr[right]);
            right++;
        }
        
        for(int i = low; i <= high; i++){
            arr[i] = temp[i - low];
        }
        
        return count;
    }
    
    
    int mergeSort(vector<int> &arr, int low, int high){
        int count = 0;
        if(low >= high) return count;
        int mid = low + (high - low)/2;
        
        count += mergeSort(arr, low, mid);
        count += mergeSort(arr, mid+1, high);
        count += merge(arr, low, mid, high);
        
        return count;
    }
  
    // Function to count inversions in the array.
    int inversionCount(vector<int> &arr) {
        return mergeSort(arr, 0, arr.size()-1);
        
    }
};


//{ Driver Code Starts.

int main() {

    int T;
    cin >> T;
    cin.ignore();
    while (T--) {
        int n;
        vector<int> a;
        string input;
        getline(cin, input);
        stringstream ss(input);
        int num;
        while (ss >> num)
            a.push_back(num);
        Solution obj;
        cout << obj.inversionCount(a) << endl;
        cout << "~" << endl;
    }

    return 0;
}

// } Driver Code Ends

File: Difficulty: Medium/Count Inversions/count-inversions.py
class Solution:
    def merge(self, low, mid, high, arr):
        temp = []
        count = 0
        left = low
        right = mid+1
        
        while left <= mid and right <= high:
            if arr[left] <= arr[right]:
                temp.append(arr[left])
                left+=1
            else:
                temp.append(arr[right])
                count += (mid - left) + 1
                right += 1
        
        while left <= mid:
            temp.append(arr[left])
            left+=1
        
        while right <= high:
            temp.append(arr[right])
            right+=1
        
        arr[low : high+1] = temp
        
        return count
    
    def mergeSort(self, low, high, arr):
        count = 0
        if low >= high:
            return count
            
        mid = (low + high)//2
        
        count += self.mergeSort(low, mid, arr)
        count += self.mergeSort(mid+1, high, arr)
        count += self.merge(low, mid, high, arr)
        
        return count
    
    def inversionCount(self, arr):
        n =  len(arr)
        return self.mergeSort(0, n-1, arr)